tests/test_tga.py

```diff
@@ -171,6 +172,69 @@ def test_prefix_tokens(self):
                 PREFIX_TEXT
             ), f"[{beam}] does not start with [{PREFIX_TEXT}]"
 
+    def test_token_level_loss_logging(self):
+        """
+        Test functionality of token level probability + ranking logging.
+
+        Regression for all inference types: 'beam', 'greedy', 'topk', 'nucleus',
+        'delayedbeam'
+        """
+        inference_types = ['beam', 'greedy', 'topk', 'nucleus', 'delayedbeam']
+        gold_data = {
+            'beam': {
+                'text_token_info': [
+                    ('__start__', 0.0, 1.0),
+                    ('5', -2.5510462364763953e-05, 1.0),
+                    ('__end__', -1.1920922133867862e-06, 3.0),
+                ],
+                'extra_args': ['--beam-size', '3'],
+            },
+            'greedy': {
+                'text_token_info': [
+                    ('__start__', 0.0, 1.0),
+                    ('5', -2.5510462364763953e-05, 1.0),
+                    ('__end__', -1.1920922133867862e-06, 1.0),
+                ],
+                'extra_args': [],
+            },
+            # sampling based token selection will produce non-deterministic output
+            'topk': {'extra_args': ['--topk', '2']},
```

Comment:
can we try these methods with beam_size > 1 as well

Reply:
Sorry if I'm misunderstanding but how can we have multiple beams in vanilla topk sampling or vanilla nucleus sampling?

Reply:
Emily and I discussed this in a meeting. The answer is no, you cannot, but ParlAI is written to allow for multiple beams in topk sampling and nucleus sampling. When there are multiple beams for either of these methods, the methods will decode sequences independently along each beam. You can think of beam size as a kind of batch size. It cannot be set for greedy decoding. This makes sense because greedy decoding always produces the same thing.